filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,524 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/common/centroid.h>
+#include <pcl/common/common.h>
+#include <pcl/filters/experimental/transform_filter.h>
+
+#include <boost/optional.hpp> // std::optional for C++17
+
+#include <unordered_map>
+
+// For testing protected method
+class VoxelGridProtectedMethods_GridFilters_Test;
+
+namespace pcl {
+namespace experimental {
+
+template <typename PointT>
+struct Voxel {
+  // TODO: find if there is better way than storing two centroid types
+  CentroidPoint<PointT> centroid;
+  Eigen::Vector4f coord_centroid = Eigen::Vector4f::Zero();
```

Comment:
Why do we need 2 of these?

Reply:
There are two ways to compute the centroids: average coordinate only / average all the fields if possible

https://github.com/PointCloudLibrary/pcl/blob/cd3b3f95441746ce5c7ce1116768d2ee6b40d996/filters/include/pcl/filters/impl/voxel_grid.hpp#L404-L423

I'm looking for some way to avoid having two centroids

Reply:
You could use a union or boost/std::variant there

Reply:
I tried `boost::variant` but it adds some unnecessary code and I think using PCL's own stuff is probably better

I found [`point.getVector4fMap()`](https://github.com/PointCloudLibrary/pcl/blob/e6db18c04b76857e5d975e4fd33987460eab66c5/common/include/pcl/impl/point_types.hpp#L252) will only return the 4DoF coordinate. Combine it with [`pcl::CentroidPoint`](https://github.com/PointCloudLibrary/pcl/blob/e6db18c04b76857e5d975e4fd33987460eab66c5/common/include/pcl/common/centroid.h#L1022) which is also used in the original code, we can achieve the same result

@kunaltyagi  what do you think about this approch?

Reply:
I tested using `union` of `CenterPoint` & `Array4f` and `CenterPoint` + `getVector4fMap`, the performance of the latter seems already quite good

## Benchmarks

### My machine
```
Run on (8 X 2800 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 6144 KiB (x1)
  L4 Unified 131072 KiB (x1)
Load Average: 1.43, 1.44, 2.27
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
```
### `union`
```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_VoxelGrid_milk                    10.8 ms         10.8 ms           64
BM_ApproximateVoxelGrid_milk         7.39 ms         7.35 ms           93
BM_Experimental_VoxelGrid_milk       10.9 ms         10.8 ms           69
BM_VoxelGrid_mug                     8.59 ms         8.51 ms           83
BM_ApproximateVoxelGrid_mug          6.90 ms         6.84 ms          102
BM_Experimental_VoxelGrid_mug        7.86 ms         7.79 ms           92
```

### `CenterPoint` + `getVector4fMap`

```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_VoxelGrid_milk                    10.9 ms         10.9 ms           64
BM_ApproximateVoxelGrid_milk         7.42 ms         7.42 ms           92
BM_Experimental_VoxelGrid_milk       8.34 ms         8.34 ms           86
BM_VoxelGrid_mug                     8.61 ms         8.60 ms           81
BM_ApproximateVoxelGrid_mug          6.79 ms         6.79 ms          103
BM_Experimental_VoxelGrid_mug        6.36 ms         6.36 ms          110
```

Reply:
@mvieth I also tested other approaches. 

`using Grid = typename std::unordered_map<std::size_t, CentroidPoint<PointT>>;`

Beside `CentroidPoint`, I tried `struct` with `Array4f` and `size_t`; `vector<size_t>` for storing point index, and construct `CentroidPoint` one by one as in the original `VoxelGrid`. But the benchmark result show `CenterPoint` + `getVector4fMap` is the notable faster than the other way I tried

Reply:
I benchmarked both `union` and `boost::variant` with default `downsample_all_data=false` (use `CentroidPoint` instead of `Vector4f`). In addition, I implement `CentroidPoint::clear()` to avoid constructing and destructing voxel objects.

Now the runtime of `office` is close to the original `VoxelGrid`

## Benchmark

### `union`
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_VoxelGrid_5people                    14.9 ms         14.9 ms           47
BM_ApproximateVoxelGrid_5people         8.34 ms         8.33 ms           83
BM_Experimental_VoxelGrid_5people       11.4 ms         11.4 ms           60
BM_VoxelGrid_office                     14.3 ms         14.3 ms           48
BM_ApproximateVoxelGrid_office          9.22 ms         9.22 ms           76
BM_Experimental_VoxelGrid_office        14.7 ms         14.7 ms           46
BM_VoxelGrid_milk                       7.86 ms         7.85 ms           89
BM_ApproximateVoxelGrid_milk            5.48 ms         5.47 ms          128
BM_Experimental_VoxelGrid_milk          5.20 ms         5.20 ms          107
BM_VoxelGrid_mug                        6.34 ms         6.34 ms          112
BM_ApproximateVoxelGrid_mug             5.04 ms         5.04 ms          136
BM_Experimental_VoxelGrid_mug           4.31 ms         4.31 ms          162

```

### `boost::variant`
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_VoxelGrid_5people                    13.6 ms         13.6 ms           52
BM_ApproximateVoxelGrid_5people         8.00 ms         7.99 ms           85
BM_Experimental_VoxelGrid_5people       11.6 ms         11.6 ms           61
BM_VoxelGrid_office                     13.5 ms         13.5 ms           50
BM_ApproximateVoxelGrid_office          8.49 ms         8.47 ms           82
BM_Experimental_VoxelGrid_office        15.8 ms         15.8 ms           43
BM_VoxelGrid_milk                       8.27 ms         8.27 ms           84
BM_ApproximateVoxelGrid_milk            5.97 ms         5.96 ms          120
BM_Experimental_VoxelGrid_milk          6.19 ms         6.19 ms          109
BM_VoxelGrid_mug                        6.58 ms         6.57 ms          105
BM_ApproximateVoxelGrid_mug             5.45 ms         5.44 ms          128
BM_Experimental_VoxelGrid_mug           5.19 ms         5.19 ms          134
```

Reply:
@mvieth I agree with you that we shoudn't use `CentroidPoint` in both case, although the benchmark can't reflect the difference because we're benchmarking clouds with same point type.

After benchmarking above, I'll switch to `union`

Reply:
Could you also compare the use of `unordered_map::clear()` and `for_each + clear`?

Reply:
`for_each + clear` is the `union` above

### `unordered_map::clear()`
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_VoxelGrid_5people                    13.8 ms         13.8 ms           51
BM_ApproximateVoxelGrid_5people         7.88 ms         7.87 ms           86
BM_Experimental_VoxelGrid_5people       25.1 ms         25.1 ms           32
BM_VoxelGrid_office                     13.8 ms         13.8 ms           51
BM_ApproximateVoxelGrid_office          8.40 ms         8.39 ms           84
BM_Experimental_VoxelGrid_office        38.7 ms         38.7 ms           21
BM_VoxelGrid_milk                       8.95 ms         8.94 ms           75
BM_ApproximateVoxelGrid_milk            6.61 ms         6.61 ms          110
BM_Experimental_VoxelGrid_milk          7.99 ms         7.98 ms           87
BM_VoxelGrid_mug                        7.13 ms         7.13 ms           95
BM_ApproximateVoxelGrid_mug             5.74 ms         5.74 ms          122
BM_Experimental_VoxelGrid_mug           5.40 ms         5.40 ms          128
```

Reply:
`clear` is just awful. Don't know why 😆 Let's stick with union and the new clear function